File: infrastructure/db-core/src/main/java/com/nmh/commerce/coupon/CouponStockEntity.java
package com.nmh.commerce.coupon;

import com.nmh.commerce.BaseEntity;
import com.nmh.commerce.domain.Quantity;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.NoArgsConstructor;

@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity
public class CouponStockEntity {
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private Long couponId;
    private Quantity remainingQuantity;
    @Version
    private Long version;

    @Builder
    private CouponStockEntity(Long id, Long couponId, Quantity remainingQuantity, Long version) {
        this.id = id;
        this.couponId = couponId;
        this.remainingQuantity = remainingQuantity;
        this.version = version;
    }

    public static CouponStockEntity from(CouponStock couponStock) {
        return CouponStockEntity.builder()
            .id(couponStock.id)
            .couponId(couponStock.couponId)
            .remainingQuantity(couponStock.remainingQuantity)
            .version(couponStock.version)
            .build();
    }

    public CouponStock toDomain() {
        return new CouponStock(id, couponId, remainingQuantity, version);
    }
}

File: infrastructure/db-core/src/main/kotlin/com/nmh/commerce/coupon/CouponStockEntity.kt
package com.nmh.commerce.coupon

import com.nmh.commerce.BaseEntity
import com.nmh.commerce.domain.Quantity
import com.nmh.commerce.utils.QuantityConverter
import jakarta.persistence.Convert
import jakarta.persistence.Entity
import jakarta.persistence.Version

@Entity
class CouponStockEntity private constructor(
    override val id: Long = 0,
    private val couponId: Long,
    @field:Convert(converter = QuantityConverter::class)
    private val remainingQuantity: Quantity,
    @field:Version
    private val version: Long?,
) : BaseEntity<Long>() {
    fun toDomain(): CouponStock = CouponStock(id, couponId, remainingQuantity, version)

    companion object {
        fun from(couponStock: CouponStock): CouponStockEntity =
            CouponStockEntity(
                couponStock.id,
                couponStock.couponId,
                couponStock.remainingQuantity,
                couponStock.version,
            )
    }
}
